packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -50,20 +50,21 @@ function warnForPluginsWithoutNodes(state: IGatsbyState, nodes: Node[]): void {
 function getStaleNodes(state: IGatsbyState, nodes: Node[]): Node[] {
   return nodes.filter(node => {
     let rootNode = node
+    let next: Node | undefined = undefined
+
     let whileCount = 0
-    while (
-      rootNode.parent &&
-      getNode(rootNode.parent) !== undefined &&
-      whileCount < 101
-    ) {
-      rootNode = getNode(rootNode.parent)
-      whileCount += 1
-      if (whileCount > 100) {
-        console.log(
-          `It looks like you have a node that's set its parent as itself`,
-          rootNode
-        )
+    do {
+      next = rootNode.parent ? getNode(rootNode.parent) : undefined
+      if (next) {
+        rootNode = next
       }
+    } while (next && ++whileCount < 101)
+
+    if (whileCount > 100) {
+      console.log(
```

Comment:
This should be a reporter.warn (though you can't just pass the node in as an arg)

Reply:
I didn't change this, only moved it. Don't want to change more than required for TS migrations.